### PR TITLE
[RJ3D-31] Convert DTE to Typescript

### PR DESCRIPTION
**Pull request by ghost**
The first commit is the major overhaul

The rest is small fixes and optimisations to make the code clearer

### CODE REVIEW

File: src/components/Model3D.tsx [hunk 1]
```diff
@@ -136,71 +162,67 @@ class Model3D extends React.Component {
...
   };
 
-  onReady = (viewer, model) => {
+  onReady = (viewer: Cognite3DViewer, model: Cognite3DModel) => {
     this.viewer = viewer;
     this.model = model;
     // Temp disable screen space culling
+    // TODO, this is not allowed
+    // @ts-ignore
     model._screenRatioLimit = 0.0;
```
Comment (dragly) [inline src/components/Model3D.tsx]: Why is this here in the first place?

File: src/components/Model3D.tsx [hunk 2]
```diff
@@ -1,29 +1,65 @@
...
+  static defaultProps = {
+    assetMappings: { byNodeId: {}, byAssetId: {} },
+    nodeId: undefined,
+    cache: undefined
+  };
+  readonly state: Readonly<State> = { hasWarnedAboutNode: {} };
 
-  componentDidUpdate(prevProps, prevState) {
+  currentColoredNodes: number[] = [];
+  model?: Cognite3DModel = undefined;
```
Comment (dragly) [inline src/components/Model3D.tsx]: Any way we could rewrite this class to be guaranteed to have a model and viewer? I.e. by passing them to the constructor or similar.

File: src/containers/AddTypes.tsx [hunk 3]
```diff
@@ -27,10 +37,8 @@ class AddTypes extends React.Component {
   };
 
   render() {
-    const existingTypeIds = this.props.asset.types.map(type => type.id);
-    const nonUsedTypes = this.props.types.filter(
-      type => existingTypeIds.indexOf(type.id) === -1
-    );
+    const existingTypeIds = this.props.asset.types ? this.props.asset.types.map(type => type.id) : [];
```
Comment (dragly) [inline src/containers/AddTypes.tsx]: Is it not allowed to do a map on an empty array?

File: src/containers/AssetDrawer.tsx [hunk 4]
```diff
@@ -52,9 +36,36 @@ const HeaderWithButton = styled.div`
...
+  events: EventsAndTypes;
+  threed: ThreeDState;
+  types: TypesState;
+  width: number;
+};
+
+type State = {
+  showAddTimeseries?: boolean;
+  asset?: ExtendedAsset;
+  showAddTypes?: boolean;
```
Comment (dragly) [inline src/containers/AssetDrawer.tsx]: Why are these undefined by default?

File: src/components/Model3D.tsx [hunk 5]
```diff
@@ -1,29 +1,65 @@
+// TODO is this ideal?
+declare global {
+  interface Window {
+    viewer: Cognite3DViewer;
+    THREE: any;
+    model: any;
```
Comment (dragly) [inline src/components/Model3D.tsx]: Is this not of type Model3D?

**Review by dragly — APPROVED** [review]
Only did a superficial review since this is a conversion, but it looks very good overall. Just a few minor questions and comments.

File: src/components/Model3D.tsx (hunk 1 above)
Comment (dlcognite) [reply src/components/Model3D.tsx]: I think i'll figure this out later, it so far seems to have no effect

File: src/components/Model3D.tsx (hunk 2 above)
Comment (dlcognite) [reply src/components/Model3D.tsx]: sadly because of React's lifecycle we cant expect anything else in the constructor and need to wait for the `onReady` callback

File: src/containers/AssetDrawer.tsx (hunk 4 above)
Comment (dlcognite) [reply src/containers/AssetDrawer.tsx]: hmm `showAddTypes` can also be a boolean i believe.
